### ISSUE TITLE
Spark integration, and examples

### ISSUE THREAD

**Issue by MLnick** [rjagerman/glint#28]
Hi @rjagerman, this project looks very interesting and I'd like to explore it a bit more. You mention Spark integration as a goal, has there been work done on that? What about example algorithms using this parameter server?

**Comment by rjagerman** [comment]
Thanks for your interest in the project! Currently the parameter server runs stand-alone outside of Spark which means you'd have to run the master and servers on a cluster as separate java processes from Spark (the scripts in the `sbin` folder can help with that). When spawning big matrices or vectors on the parameter server you get a `BigMatrix` and `BigVector` object back respectively. These objects are serializable and can safely be used within a spark function, thus running it with Spark is extremely easy:

    val matrix = glintclient.matrix[Double](100000, 1000)
    val rdd = ... // some spark RDD
    rdd.foreachPartition { case partition => 
        ... // this will get executed on your spark workers
        ... // they can safely use the matrix object
        ... // e.g. matrix.pull(Array(0, 1, 2 , 3))
    }

This project is part of my ongoing Master thesis. I currently have an implementation of LDA that uses this parameter server and Spark that scales up to 2TB of data and 1000s of topics on a moderate computing cluster. I will open source that implementation within the next month or so.

I will also write more extensive documentation going into more detail how to set up a cluster, some code examples and some tuning tips with regards to ExecutionContexts, timeouts, etc. Due to other deadlines regarding my thesis I haven't found the time for this yet.

**Comment by MLnick** [comment]
Ok, thanks. I'd like to see if I can find some time to investigate some implementations on top of this.

Do you know how it compares to https://github.com/dmlc/ps-lite in terms of performance etc? In terms of ease-of-integration with frameworks like Spark, something based on Scala/Akka seems much nicer.

**Comment by rjagerman** [comment]
I currently have not compared the performance to ps-lite (or any other parameter server for that matter), but it would be a very interesting thing to measure. If I had to guess I'd say that ps-lite is much faster. Especially due to the very early alpha-state and experimental nature of Glint compared to the very mature implementation of ps-lite. My main goal is indeed to have a parameter server that is very easily integrated with Spark. So far it seems to work well in my practical case (LDA with Collapsed Gibbs sampling) but some raw numbers like updates/sec, requests/sec, etc. would be very interesting. I'll keep you updated here when I have some real measurements.

Additionally Glint has no regard for fault tolerance (unlike ps-lite, which offers instantaneous failover), so if a server goes down the data is lost. This is something I definitely wish to address in the future but is at the moment outside the scope of the project.

**Comment by MLnick** [comment]
I've started working on some POCs for Spark integration starting with
linear models for simplicity. I will see if I can do some performance
testing at some point. Would you be willing to share some code samples you
have for Spark integration (privately if need be)?

**Comment by rjagerman** [comment]
That sounds great! :+1: 

I have just now open sourced the LDA implementation [here](https://github.com/rjagerman/glintlda) so you could take a look at that. It is based on some state-of-the-art LDA research and the many low-level optimizations, caches and buffers make the code base a bit hard to follow. I can give you some pointers:

The construction of the count-table matrix for the collapsed Gibbs sampler happens [here](https://github.com/rjagerman/glintlda/blob/master/src/main/scala/glintlda/LDAModel.scala#L133)

    val topicWordCounts = gc.matrix[Long](config.vocabularyTerms, config.topics, 2, (x,y) => CyclicPartitioner(x, y))

The solver uses Spark to map partitions of our RDD to resampled partitions (effectively doing an LDA iteration) [here](https://github.com/rjagerman/glintlda/blob/master/src/main/scala/glintlda/Solver.scala#L255)

    rdd = rdd.mapPartitionsWithIndex { case (id, it) =>
        val s = solver(model, id)
        val partitionSamples = it.toArray
        s.fit(partitionSamples, t)
        partitionSamples.toIterator
    }

The `s.fit(...)` function internally uses the earlier constructed matrix on the parameter servers to perform an actual iteration of the algorithm and updates the counts on the matrix accordingly. It iterates over the matrix on the parameter servers [here](https://github.com/rjagerman/glintlda/blob/master/src/main/scala/glintlda/mh/MHSolver.scala#L44) and updates counts on the parameter server [here](https://github.com/rjagerman/glintlda/blob/master/src/main/scala/glintlda/mh/MHSolver.scala#L245) and [here](https://github.com/rjagerman/glintlda/blob/master/src/main/scala/glintlda/mh/MHSolver.scala#L275).

We make extensive use of buffers for performance reasons, so this can obfuscate the code quite a bit. The locks that you see in the code act as a back pressure mechanism that limit the number of open requests to the parameter servers. This is necessary since the code is so fast that it could easily flood the parameter server by asynchronously sending requests.


I will be finished with my thesis in about 2 weeks, after which I'll have some more time to create some minimal examples that are much easier to read and understand. In the mean time I hope this helps! :-)

**Comment by MLnick** [comment]
Great, I will take a deeper look at that code. The general approach is
along the lines of what my approach is (i.e. run iterations within
mapPartitions).

**Comment by rjagerman** [comment]
Also another hint that could be helpful: In my glint configuration I increased the Akka frame size and heart beat timeouts. This allows me to send much larger pull and push requests. My configuration file looks like this:

```
glint.master.host   = "127.0.0.1"
glint.master.port   = 13370
glint {
  server.akka.loglevel = "INFO"
  server.akka.stdout-loglevel = "INFO"
  client.akka.loglevel = "ERROR"
  client.akka.stdout-loglevel = "ERROR"
  master.akka.loglevel = "INFO"
  master.akka.stdout-loglevel = "INFO"
  master.akka.remote.log-remote-lifecycle-events = on
  server.akka.remote.log-remote-lifecycle-events = off
  client.akka.remote.log-remote-lifecycle-events = on
  client.timeout = 30 s

  master.akka.remote.transport-failure-detector.acceptable-heartbeat-pause = 120 s
  server.akka.remote.transport-failure-detector.acceptable-heartbeat-pause = 120 s
  client.akka.remote.transport-failure-detector.acceptable-heartbeat-pause = 120 s
  master.akka.remote.watch-failure-detector.acceptable-heartbeat-pause = 120 s
  server.akka.remote.watch-failure-detector.acceptable-heartbeat-pause = 120 s
  client.akka.remote.watch-failure-detector.acceptable-heartbeat-pause = 120 s

  server.akka.remote.netty.tcp.maximum-frame-size = 10240000b
  client.akka.remote.netty.tcp.maximum-frame-size = 10240000b
  server.akka.remote.netty.tcp.send-buffer-size = 20480000b
  client.akka.remote.netty.tcp.send-buffer-size = 20480000b
  server.akka.remote.netty.tcp.receive-buffer-size = 20480000b
  client.akka.remote.netty.tcp.receive-buffer-size = 20480000b
}
```

**Comment by rjagerman** [comment]
I've added slightly more comprehensive documentation at http://rjagerman.github.io/glint/. This might be of use to you as it includes a short section on spark integration and serialization (see the getting started guide).

I'm also currently working on getting some benchmarks for common tasks (e.g. logistic regression, SVMs, regression, all-reduce, etc.) so we can compare glint against spark and other frameworks. In time this will produce some more example code for these common tasks.

**Comment by MLnick** [comment]
Thanks Rolf! I have been swamped with Spark 2.0 work, but I'd like to get
back to some PoC work with glint after that is done

**Comment by codlife** [comment]
Hi @rjagerman ,After the look of your code ,I think it's clear to use the your API  in the spark map or mapPartitions operations ,But how about the spark aggregate,I find it's a little hard to use your API,such as I want to implement Gradient Descent. Thanks!
